ADRs/0003-Pipeline_API_Rewrite.md

```diff
@@ -0,0 +1,435 @@
+      `BaseStepRunner.transform(Object...)` and `.transform(Object[][])` just unwrap/convert them and calls `transform(Record[])`
+      internally anyway
+    - Note 1: _none_ of these are suitable for sequences (without ugly hacks like embedding sequence as an NDArray)
+    - Note 2: Even if we added DataVec `SequenceRecord`/`List<List<Writable>>` support for sequences, there will still be
+      no way to mix sequence and non-sequence data (again, without ugly hacks), or have sequences of different lengths
+* A DataVec-based API is a significant blocker for adding other programming languages - what happens if we want a pipeline
+  step based on C#, C++, Swift, JavaScript - there's no DataVec API here
+* When communicating between languages or VMs (if a pipeline was to be split between processes/VMs as in a microservice),
+  a DataVec-based API adds significant serialization overheads
+* No support for dynamic schemas or optional return values (for example in classification: optionally return probabilities or not based on user request) 
+* Writing custom (Java) pipeline steps requires users to be familiar with DataVec's APIs (as well as conversion to/from INDArray, etc)
+* Input.DataFormat and Output.DataFormat contain language-specific features (NUMPY, ND4J)
+* The Input.DataFormat and Output.DataFormat are restrictive: for example, there's no nice/easy way to return a String,
+  an arbitrary byte[], or an image (for a segmentation model), at least without hacks like wrapping them in and NDArray
+* Multi-output networks are not really supported. For example, what if I have a multi-task classification + regression model?
+  A single PredictionType setting doesn't work here 
+* Invalid combinations of configuration are easy to produce - what does YOLO prediction type on a JSON or ARROW Output.DataFormat
+  mean?
+    - Similarly, the Python client is not able to handle a Output.DataFormat of ND4J [Link](https://github.com/KonduitAI/konduit-serving/blob/825a10d7ba356b502a6a743c0532b0828546a365/python/konduit/client.py#L217-L218)
+* No way to return metadata to users
+
+
+In short, there are issues with usability, performance, maintainability and support for some use cases in the current design. 
+
+### Requirements for a new API and data formats
+
+Suppose we redesign the API for handling data. This will have consequences throughout the codebase, which changes impacting:
+* Pipeline and PipelineStep APIs
+* YAML/JSON Configuration (serving, client, and steps)
+* REST Endpoints
+* Client API
+
+**Requirements**
+
+* Extensible to other languages - can in principle allow interop with languages we might one day need to deploy, such as C#, C++, JavaScript, Matlab, etc
+* Needs to support efficient serialization/deserialization (ideally, zero copy where possible)
+* Suitable for a microservices type split of pipeline steps (i.e., some steps run in different processes/containers)
+* Efficient for 'monolithic' deployment options
```

Comment:
> Can we clarify where we pay for this?

Right now: we don't have that cost (and that's a good thing), because everything is monolithic (run in the one JVM). My point is we don't want to introduce a design/data structure that requires any sort of serialization or overhead for monolithic deployments; we should only pay a serialization cost if there's IPC (like if the pipeline is split between multiple containers/devices for some reason).